### ISSUE TITLE
ko run does not delete pod upon interrupt

### ISSUE THREAD

**Issue by ahmetb** [google/ko#304]
Hitting Ctrl+C after pod has started leaks the pod:

```
Flag --generator has been deprecated, has no effect and will be removed in the future.
If you don't see a command prompt, try pressing enter.
<<POD RUNS FINE>>

^C2021/02/03 10:46:55 error executing "kubectl run": signal: interrupt
```

```
NAME                                      READY   STATUS    RESTARTS   AGE
cancel-10aec35353f9c4096a71c38654c3d402   1/1     Running   0          57s
```

from the speed of ctrl+c it seems it's not even calling into kubectl.

I compiled from HEAD yesterday. 82cabb40bae57

**Comment by ImJasonH** [comment]
`ko run` does a `ko publish` then `kubectl run --attach`, which itself is basically a `kubectl run` then `kubectl attach`. It doesn't look like `kubectl attach` handles ctrl+c by killing the container either, so I think this might be working as intended, in need of better documentation?

**Comment by jonjohnsonjr** [comment]
I don't think we document `ko run` anywhere, and maybe we should hide it... it's a bit janky.

**Comment by jonjohnsonjr** [comment]
> Flag --generator has been deprecated, has no effect and will be removed in the future.

hmmm

**Comment by ImJasonH** [comment]
> I don't think we document `ko run` anywhere, and maybe we should hide it... it's a bit janky.

I'd be fine with that. I'm not sure if @mattmoor uses it for anything. It was added way back in  https://github.com/google/go-containerregistry/pull/331

**Comment by ahmetb** [comment]
At least we can add `--rm` for client-side deletion upon passing sigterm to kubectl.

**Comment by jonjohnsonjr** [comment]
https://github.com/google/ko/blob/82cabb40bae577ce3bc016e5939fd85889538e8b/pkg/commands/run.go#L109

lol

> At least we can add --rm for client-side deletion upon passing sigterm to kubectl.

https://github.com/google/ko/blob/82cabb40bae577ce3bc016e5939fd85889538e8b/pkg/commands/run.go#L100

🤷 

Not sure?

**Comment by ahmetb** [comment]
I think the problem is using `exec.CommandContext(ctx, "kubectl", argv...)` which gives you no control over how the underlying program is terminated.

If we do exec.Command() followed by `Start()`, then have a `select` choosing between result of `Wait()` routine and `<-ctx.Done()`, we can have fine control over the underlying process (we can signal `Kill(os.Interrupt)` to the started pid)?

**Comment by jonjohnsonjr** [comment]
> If we do exec.Command() followed by Start(), then have a select choosing between result of Wait() routine and <-ctx.Done(), we can have fine control over the underlying process 

Makes sense to me. I think I threaded this ctx through somewhat recently because before interrupting just did nothing, but I'm open to more explicit cleanup.

**Comment by ahmetb** [comment]
Actually I think my assumption was wrong: `kubectl run` command does not pass on the signals. However if you run it with `-t`, it forwards your keystrokes such as <kbd>Ctrl</kbd>+<kbd>C</kbd> to the container.

I think the solution here would be to convert the termination signal into a keystroke we send over PTY to the kubectl’s stdin. Overkill?

**Comment by ahmetb** [comment]
Actually here's a better idea, instead of dealing with low-level APIs to send keystrokes to kubectl, we can just [`syscall.Exec`](https://golang.org/pkg/syscall/#Exec) on darwin/linux and replace `ko` process entirely with `kubectl`. We can't do execve on Windows, but it's a fair trade-off.

**Comment by jonjohnsonjr** [comment]
> replace ko process entirely with kubectl

Neat!
